tracing/src/main/java/com/palantir/tracing/Trace.java

```diff
@@ -28,62 +34,214 @@
-    }
+    abstract void push(OpenSpan span);
+
+    abstract Optional<OpenSpan> top();
+
+    abstract Optional<OpenSpan> pop();
+
+    abstract boolean isEmpty();
 
     /**
      * True iff the spans of this trace are to be observed by {@link SpanObserver span obververs} upon {@link
      * Tracer#completeSpan span completion}.
      */
-    boolean isObservable() {
-        return isObservable;
-    }
+    abstract boolean isObservable();
 
     /**
      * The globally unique non-empty identifier for this call trace.
      */
-    String getTraceId() {
+    final String getTraceId() {
         return traceId;
     }
 
     /** Returns a copy of this Trace which can be independently mutated. */
-    Trace deepCopy() {
-        return new Trace(new ArrayDeque<>(stack), isObservable, traceId);
+    abstract Trace deepCopy();
+
+    static Trace of(boolean isObservable, String traceId) {
+        return isObservable ? new Sampled(traceId) : new Unsampled(traceId);
+    }
+
+    private static final class Sampled extends Trace {
+
+        private final Deque<OpenSpan> stack;
+
+        private Sampled(ArrayDeque<OpenSpan> stack, String traceId) {
+            super(traceId);
+            this.stack = stack;
+        }
+
+        private Sampled(String traceId) {
+            this(new ArrayDeque<>(), traceId);
+        }
+
+        @Override
+        @SuppressWarnings("ResultOfMethodCallIgnored") // Sampled traces cannot optimize this path
+        void fastStartSpan(String operation, String parentSpanId, SpanType type) {
+            startSpan(operation, parentSpanId, type);
+        }
+
+        @Override
+        @SuppressWarnings("ResultOfMethodCallIgnored") // Sampled traces cannot optimize this path
+        void fastStartSpan(String operation, SpanType type) {
+            startSpan(operation, type);
+        }
+
+        @Override
+        void push(OpenSpan span) {
+            stack.push(span);
+        }
+
+        @Override
+        Optional<OpenSpan> top() {
+            return stack.isEmpty() ? Optional.empty() : Optional.of(stack.peekFirst());
+        }
+
+        @Override
+        Optional<OpenSpan> pop() {
+            return stack.isEmpty() ? Optional.empty() : Optional.of(stack.pop());
+        }
+
+        @Override
+        boolean isEmpty() {
+            return stack.isEmpty();
+        }
+
+        @Override
+        boolean isObservable() {
+            return true;
+        }
+
+        @Override
+        Trace deepCopy() {
+            return new Sampled(new ArrayDeque<>(stack), getTraceId());
+        }
+
+        @Override
+        public String toString() {
+            return "Trace{stack=" + stack + ", isObservable=true, traceId='" + getTraceId() + "'}";
+        }
     }
 
-    @Override
-    public String toString() {
-        return "Trace{stack=" + stack + ", isObservable=" + isObservable + ", traceId='" + traceId + "'}";
+    private static final class Unsampled extends Trace {
+        private int depth;
```

Comment:
could we rename this `s/depth/numberOfSpans/`, maybe also a little comment to explain that the whole point of an unsampled trace is that it'll never notify SpanObservers, we don't need to store any of the span names that users provide